File: CSharp/SpaceVIL/SpaceVIL/Engine/ActionManager.cs
using System;
using System.Threading;
using System.Collections.Generic;
using System.Collections.Concurrent;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal sealed class ActionManager
    {
        ConcurrentQueue<EventTask> StackEvents = new ConcurrentQueue<EventTask>();

        internal ManualResetEventSlim Execute = new ManualResetEventSlim(false);
        WindowLayout _handler;
        bool _stoped;
        private Object managerLock = new Object();

        internal ActionManager(WindowLayout wnd)
        {
            _handler = wnd;
        }
        internal void StartManager()
        {
            _stoped = false;
            while (!_stoped)
            {
                Execute.Wait();
                Monitor.Enter(managerLock);
                try
                {
                    ExecuteActions();
                }
                finally
                {
                    Execute.Reset();
                    Monitor.Exit(managerLock);
                }
            }
        }

        internal void AddTask(EventTask task)
        {
            Monitor.Enter(managerLock);
            try
            {
                StackEvents.Enqueue(task);
            }
            finally
            {
                Monitor.Exit(managerLock);
            }
        }

        internal void StopManager()
        {
            _stoped = true;
        }

        internal void ExecuteActions()
        {
            if (StackEvents.Count == 0)
                return;

            while (StackEvents.Count > 0)
            {
                EventTask tmp = null;
                if (StackEvents.TryDequeue(out tmp))
                {
                    try
                    {
                        ExecuteAction(tmp);
                    }
                    catch (System.Exception ex)
                    {
                        Console.WriteLine(tmp.Item.GetItemName() + " " + tmp.Action + " " + ex.ToString());
                        Console.WriteLine(ex.StackTrace);
                    }
                }
            }
        }
        private void ExecuteAction(EventTask task)
        {
            switch (task.Action)
            {
                case InputEventType.MouseRelease:
                    InvokeMouseClickEvent(task.Item, task.Args as MouseArgs);
                    break;
                case InputEventType.MousePress:
                    InvokeMousePressEvent(task.Item, task.Args as MouseArgs);
                    break;
                case InputEventType.MouseDoubleClick:
                    InvokeMouseDoubleClickEvent(task.Item, task.Args as MouseArgs);
                    break;
                case InputEventType.MouseHover:
                    InvokeMouseHoverEvent(task.Item, task.Args as MouseArgs);
                    break;
                case InputEventType.MouseDrag:
                    InvokeMouseDragEvent(task.Item, task.Args as MouseArgs);
                    break;
                case InputEventType.FocusGet:
                    InvokeFocusGetEvent(task.Item);
                    break;
                case InputEventType.KeyPress:
                    InvokeKeyPressEvent(task.Item, task.Args as KeyArgs);
                    break;
                case InputEventType.KeyRelease:
                    InvokeKeyReleaseEvent(task.Item, task.Args as KeyArgs);
                    break;
                case InputEventType.TextInput:
                    InvokeTextInputEvent(task.Item, task.Args as TextInputArgs);
                    break;
                case InputEventType.WindowDrop:
                    InvokeDropEvent(task.Item, task.Args as DropArgs);
                    break;
                default:
                    break;
            }
        }

        //common events
        private void InvokeFocusGetEvent(Prototype sender)
        {
            sender.EventFocusGet?.Invoke(sender);
        }
        private void InvokeFocusLostEvent(Prototype sender)
        {
            sender.EventFocusLost?.Invoke(sender);
        }
        private void InvokeResizedEvent(Prototype sender)
        {
            sender.EventResize?.Invoke(sender);
        }
        private void InvokeDestroyedEvent(Prototype sender)
        {
            sender.EventDestroy?.Invoke(sender);
        }

        //mouse input
        private void InvokeMouseClickEvent(Prototype sender, MouseArgs args)
        {
            sender.EventMouseClick?.Invoke(sender, args);
        }
        private void InvokeMouseHoverEvent(Prototype sender, MouseArgs args)
        {
            sender.EventMouseHover?.Invoke(sender, args);
        }
        private void InvokeMousePressEvent(Prototype sender, MouseArgs args)
        {
            sender.EventMousePress?.Invoke(sender, args);
        }
        private void InvokeMouseDoubleClickEvent(Prototype sender, MouseArgs args)
        {
            if (sender.EventMouseDoubleClick == null
                || sender.EventMouseDoubleClick.GetInvocationList().Length == 0)
                InvokeMouseClickEvent(sender, args);
            else
                sender.EventMouseDoubleClick?.Invoke(sender, args);
        }
        private void InvokeMouseDragEvent(Prototype sender, MouseArgs args)
        {
            sender.EventMouseDrag?.Invoke(sender, args);
        }
        private void InvokeMouseDropEvent(Prototype sender, MouseArgs args)
        {
            sender.EventMouseDrag?.Invoke(sender, args);
        }
        private void InvokeMouseScrollUpEvent(Prototype sender, MouseArgs args)
        {
            sender.EventScrollUp?.Invoke(sender, args);
        }
        private void InvokeMouseScrollDownEvent(Prototype sender, MouseArgs args)
        {
            sender.EventScrollDown?.Invoke(sender, args);
        }

        //keyboard input
        private void InvokeKeyPressEvent(Prototype sender, KeyArgs args)
        {
            sender.EventKeyPress?.Invoke(sender, args);
        }
        private void InvokeKeyReleaseEvent(Prototype sender, KeyArgs args)
        {
            sender.EventKeyRelease?.Invoke(sender, args);
        }
        private void InvokeTextInputEvent(Prototype sender, TextInputArgs args)
        {
            sender.EventTextInput?.Invoke(sender, args);
        }

        //window events
        private void InvokeDropEvent(Prototype sender, DropArgs args)
        {
            WContainer window = sender as WContainer;
            window?.EventDrop?.Invoke(sender, args);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/ActionManager.java
package com.spvessel;

import com.spvessel.Core.KeyArgs;
import com.spvessel.Core.MouseArgs;
import com.spvessel.Core.TextInputArgs;

import java.util.concurrent.*;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

class ActionManager {
    public ConcurrentLinkedQueue<EventTask> stackEvents = new ConcurrentLinkedQueue<EventTask>();

    public ManualResetEvent execute = new ManualResetEvent(false);
    private Lock managerLock = new ReentrantLock();
    WindowLayout _handler;
    boolean _stoped;

    public ActionManager(WindowLayout wnd) {
        _handler = wnd;
    }

    public void startManager() {
        _stoped = false;
        while (!_stoped) {
            try {
                execute.waitOne();
            } catch (InterruptedException e) {
            }
            managerLock.lock();
            try {
                executeActions();
            } finally {
                execute.reset();
                managerLock.unlock();
            }
        }
    }

    public void addTask(EventTask task) {
        managerLock.lock();
        try {
            stackEvents.add(task);
        } finally {
            managerLock.unlock();
        }
    }

    public void stopManager() {
        _stoped = true;
    }

    protected void executeActions() {
        if (stackEvents.size() == 0)
            return;

        while (stackEvents.size() > 0) {
            EventTask tmp = stackEvents.poll();
            if (tmp != null) {
                executeAction(tmp);
            }
        }
    }

    private void executeAction(EventTask task) {
        switch (task.action) {
        case MOUSE_RELEASE:
            invokeMouseClickEvent(task.item, (MouseArgs) task.args);
            break;
        case MOUSE_DOUBLE_CLICK:
            invokeMouseDoubleClickEvent(task.item, (MouseArgs) task.args);
            break;
        case MOUSE_PRESS:
            invokeMousePressedEvent(task.item, (MouseArgs) task.args);
            break;
        case MOUSE_HOVER:
            invokeMouseHoverEvent(task.item, (MouseArgs) task.args);
            break;
        case MOUSE_DRAG:
            invokeMouseDragEvent(task.item, (MouseArgs) task.args);
            break;
        case FOCUS_GET:
            invokeFocusGetEvent(task.item);
            break;
        case KEY_PRESS:
            invokeKeyPressEvent(task.item, (KeyArgs) task.args);
            break;
        case KEY_RELEASE:
            invokeKeyReleaseEvent(task.item, (KeyArgs) task.args);
            break;
        case TEXT_INPUT:
            invokeTextInputEvent(task.item, (TextInputArgs) task.args);
            break;
        default:
            break;
        }
    }

    // common events
    private void invokeFocusGetEvent(Prototype sender) {
        if (sender.eventFocusGet != null)
            sender.eventFocusGet.execute(sender);
    }

    private void invokeFocusLostEvent(Prototype sender) {
        if (sender.eventFocusLost != null)
            sender.eventFocusLost.execute(sender);
    }

    private void invokeResizedEvent(Prototype sender) {
        if (sender.eventResize != null)
            sender.eventResize.execute(sender);
    }

    private void invokeDestroyedEvent(Prototype sender) {
        if (sender.eventDestroy != null)
            sender.eventDestroy.execute(sender);
    }

    // mouse input
    private void invokeMouseClickEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMouseClick != null)
            sender.eventMouseClick.execute(sender, args);
    }
    private void invokeMouseDoubleClickEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMouseDoubleClick != null)
            sender.eventMouseDoubleClick.execute(sender, args);
    }

    private void invokeMouseHoverEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMouseHover != null)
            sender.eventMouseHover.execute(sender, args);
    }

    private void invokeMousePressedEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMousePress != null)
            sender.eventMousePress.execute(sender, args);
    }

    private void invokeMouseDragEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMouseDrag != null)
            sender.eventMouseDrag.execute(sender, args);
    }

    private void invokeMouseDropEvent(Prototype sender, MouseArgs args) {
        if (sender.eventMouseDrag != null)
            sender.eventMouseDrag.execute(sender, args);
    }

    private void invokeMouseScrollUpEvent(Prototype sender, MouseArgs args) {
        if (sender.eventScrollUp != null)
            sender.eventScrollUp.execute(sender, args);
    }

    private void invokeMouseScrollDownEvent(Prototype sender, MouseArgs args) {
        if (sender.eventScrollDown != null)
            sender.eventScrollDown.execute(sender, args);
    }

    // keyboard input
    private void invokeKeyPressEvent(Prototype sender, KeyArgs args) {
        if (sender.eventKeyPress != null)
            sender.eventKeyPress.execute(sender, args);
    }

    private void invokeKeyReleaseEvent(Prototype sender, KeyArgs args) {
        if (sender.eventKeyRelease != null)
            sender.eventKeyRelease.execute(sender, args);
    }

    private void invokeTextInputEvent(Prototype sender, TextInputArgs args) {
        if (sender.eventTextInput != null)
            sender.eventTextInput.execute(sender, args);
    }
}